### PR TITLE
Warn on SSH connection for incorrect configuration

### PR DESCRIPTION
**Pull request by Gusted**
- When `setting.RepoRootPath` cannot be found(most likely due to incorrect configuration) show "Gitea: Incorrect configuration" on the client-side to help easier with debugging the problem.

### CODE REVIEW

File: cmd/serv.go [hunk 1]
```diff
@@ -296,6 +296,16 @@ func runServ(c *cli.Context) error {
 		gitcmd = exec.CommandContext(ctx, verb, repoPath)
 	}
 
+	// Check if setting.RepoRootPath exists. It could be the case that it doesn't exist, this can happen when
+	// `[repository]` `ROOT` is a relative path and $GITEA_WORK_DIR isn't passed to the SSH connection.
+	if _, err := os.Stat(setting.RepoRootPath); err != nil {
+		if os.IsNotExist(err) {
+			return fail("Incorrect configuration.",
+				"Directory %q was not found, please check if $GITEA_WORK_DIR is passed to the SSH connection or make `[repository]` `ROOT` a absolute value.",
```
Comment (delvh) [inline cmd/serv.go]:
```suggestion
				"Directory %q was not found, please check if $GITEA_WORK_DIR is passed to the SSH connection or make `[repository]` `ROOT` an absolute value.",
```

**Review by delvh — APPROVED** [review]

**Review by wxiaoguang — APPROVED** [review]
~~`fail(RepoRootPath)` would leak some server information to end users.~~

~~Since it's a rare case and shouldn't occur when Gitea works correctly, I can accept it.~~

~~A better approach is only showing the RepoRootPath in logs.~~

----

Update: only the first arg of `fail` (aka userMessage) will be returned to end user, while the second arg won't and it's for server log. Sorry for the misleading.

**Comment by lunny** [comment]
`setting.RepoRootPath` should not be leaked to client users.

File: cmd/serv.go [hunk 2]
```diff
@@ -296,6 +296,16 @@ func runServ(c *cli.Context) error {
 		gitcmd = exec.CommandContext(ctx, verb, repoPath)
 	}
 
+	// Check if setting.RepoRootPath exists. It could be the case that it doesn't exist, this can happen when
+	// `[repository]` `ROOT` is a relative path and $GITEA_WORK_DIR isn't passed to the SSH connection.
+	if _, err := os.Stat(setting.RepoRootPath); err != nil {
+		if os.IsNotExist(err) {
+			return fail("Incorrect configuration.",
+				"Directory %q was not found, please check if $GITEA_WORK_DIR is passed to the SSH connection or make `[repository]` `ROOT` an absolute value.",
+				setting.RepoRootPath)
```
Comment (Gusted) [inline cmd/serv.go]:
```suggestion
			return fail("Incorrect configuration.",
				"Directory "`[repository]` `ROOT`" was not found, please check if $GITEA_WORK_DIR is passed to the SSH connection or make `[repository]` `ROOT` an absolute value.")
```

**Comment by Gusted** [comment]
> A better approach is only showing the RepoRootPath in logs.

That's a problem... Those kinds of logs would only be show when `ENABLE_SSH_LOG` is enabled(which currently isn't mentioned in the dogs, see https://github.com/go-gitea/gitea/pull/19316).

**Review by wxiaoguang — APPROVED** [review]

File: cmd/serv.go [hunk 3]
```diff
@@ -296,6 +296,15 @@ func runServ(c *cli.Context) error {
 		gitcmd = exec.CommandContext(ctx, verb, repoPath)
 	}
 
+	// Check if setting.RepoRootPath exists. It could be the case that it doesn't exist, this can happen when
+	// `[repository]` `ROOT` is a relative path and $GITEA_WORK_DIR isn't passed to the SSH connection.
+	if _, err := os.Stat(setting.RepoRootPath); err != nil {
+		if os.IsNotExist(err) {
+			return fail("Incorrect configuration.",
+				"Directory "`[repository]` `ROOT`" was not found, please check if $GITEA_WORK_DIR is passed to the SSH connection or make `[repository]` `ROOT` an absolute value.")
```
Comment (wxiaoguang) [inline cmd/serv.go]:
```suggestion
				"Directory `[repository]` `ROOT` was not found, please check if $GITEA_WORK_DIR is passed to the SSH connection or make `[repository]` `ROOT` an absolute value.")
```

Comment (Gusted) [reply cmd/serv.go]: Ack

**Review by delvh — APPROVED** [review]

**Comment by 6543** [comment]
-> #19437

**Comment by wxiaoguang** [comment]
To correct the review before (which was incorrect):

> `fail(RepoRootPath)` would leak some server information to end users.
> 
> Since it's a rare case and shouldn't occur when Gitea works correctly, I can accept it.
> 
> A better approach is only showing the RepoRootPath in logs.

Only the first arg of `fail` (aka userMessage) will be returned to end user, while the second arg won't and it's for server log. Sorry for the misleading.

**Pull request merged**